File: ios/SwimmingCompetition/SwimmingCompetition/Model/Competition.swift
//
//  Competition.swift
//  SwimmingCompetition
//
//  Created by Aviel on 18/01/2018.
//  Copyright © 2018 Aviel. All rights reserved.
//

import Foundation

class Competition {
    
    var name:String
    var activityDate:String
    var swimmingStyle:String
    var length: String
    var numOfParticipants:String
    var id: String
    var fromAge: String
    var toAge: String
    var participants = [Participant]()
    var currentParticipants = [Participant]()
    var isDone: Bool
    
    init(json: JSON, id: String) {
        let name = json["name"] as? String
        let activityDate = json["activityDate"] as? String
        let swimmingStyle = json["swimmingStyle"] as? String
        var length = json["length"] as? String
        if length == nil {
            length = (json["length"] as? Int) != nil ? String(json["length"] as! Int) : ""
        }
        var numOfParticipants = json["numOfParticipants"] as? String
        if numOfParticipants == nil {
            numOfParticipants = (json["numOfParticipants"] as? Int) != nil ? String(json["numOfParticipants"] as! Int) : ""
        }
        let fromAge = json["fromAge"] as? String
        let toAge = json["toAge"] as? String
        let id = id
        let participants = json["participants"] as? JSON
        let currentParticipants = json["currentParticipants"] as? JSON
        var isDone:Bool = false
        if(json["isDone"] as? String == "true" || json["isDone"] as? String == "1" || json["isDone"] as? Bool == true) {
            isDone = true
        }
       
        
        self.name = name != nil ? name! : ""
        self.activityDate = activityDate != nil ? activityDate! : ""
        self.swimmingStyle = swimmingStyle != nil ? swimmingStyle! : ""
        self.length = length != nil ? length! : ""
        self.numOfParticipants = numOfParticipants != nil ? numOfParticipants! : ""
        self.fromAge = fromAge != nil ? fromAge! : ""
        self.toAge = toAge != nil ? toAge! : ""
        self.id = id
        self.participants = []
        self.currentParticipants = []
        self.isDone = isDone
        
        if participants != nil {
            for part in participants! {
                var participant : Participant!
                let data = participants![part.0] as! JSON
                participant = Participant(json: data, id: part.0)
                self.participants.append(participant)
            }
        }
        
        if currentParticipants != nil {
            for part in currentParticipants! {
                var currentParticipant: Participant!
                let data = currentParticipants![part.0] as! JSON
                currentParticipant = Participant(json: data, id: part.0)
                self.currentParticipants.append(currentParticipant)
            }
        }
    }
    
    func getId() -> String {
        return self.id
    }
    
    func getName() -> String {
        return self.name
    }
    
    func setName(name: String) {
        self.name = name
    }
    
    func getActivityDate() -> String {
        return self.activityDate
    }
    
    func setActivityDate(activityDate: String) {
        self.activityDate = activityDate
    }
    
    func getSwimmingStyle() -> String {
        return self.swimmingStyle
    }
    
    func setSwimmingStyle(swimmingStyle: String) {
        self.swimmingStyle = swimmingStyle
    }
    
    func getLength() -> String {
        return self.length
    }
    
    func setLength(length: String) {
        self.length = length
    }
    
    func getNumOfParticipants() -> String {
        return self.numOfParticipants
    }
    
    func setNumOfParticipants(numOfParticipants: String) {
        self.numOfParticipants = numOfParticipants
    }
    
    func getFromAge() -> String {
        return self.fromAge
    }
    
    func setFromAge(fromAge: String) {
        self.fromAge = fromAge
    }
    
    func getToAge() -> String {
        return self.toAge
    }
    
    func setToAge(toAge: String) {
        self.toAge = toAge
    }
    
}

File: android/app/src/main/java/com/swimming/amirl/swimmimg_competitions/Competition.java
package com.swimming.amirl.swimmimg_competitions;

import org.json.JSONException;
import org.json.JSONObject;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Iterator;

public class Competition implements Serializable {

    private String id;
    private String name;
    private String swimmingStyle;
    private String participants;
    private String currentParticipants;
    private Boolean isDone;
    private String activityDate;
    private String numOfParticipants;
    private String fromAge;
    private String toAge;
    private String length;


    public Competition(String id, String name, String activityDate, String swimmingStyle, Integer numOfParticipants, Integer fromAge, Integer toAge, Integer length) {
        this.id = id;
        this.name = name;
        this.activityDate = activityDate;
        this.numOfParticipants = String.valueOf(numOfParticipants);
        this.swimmingStyle = swimmingStyle;
        this.length = String.valueOf(length);
        this.fromAge = String.valueOf(fromAge);
        this.toAge = String.valueOf(toAge);
    }

    public Competition(String id, JSONObject data) throws JSONException {
        this.id = id;
        this.name = data.getString("name");
        this.activityDate = data.getString("activityDate");
        this.numOfParticipants = data.getString("numOfParticipants");
        this.swimmingStyle = data.getString("swimmingStyle");
        this.length = data.getString("length");
        this.fromAge = data.getString("fromAge");
        this.toAge = data.getString("toAge");

        if (data.has("participants")) {
            this.participants = data.getString("participants");
        }

        if (data.has("currentParticipants")) {
            this.currentParticipants = data.getString("currentParticipants");
        }
    }

    public Competition(JSONObject data) throws JSONException {
        this(data.getString("id"), data);
    }

    public JSONObject getJSON_Object() throws JSONException {
        JSONObject data = new JSONObject();

        data.put("id", this.id);
        data.put("name", this.name);
        data.put("participants", this.participants);
        data.put("currentParticipants", this.currentParticipants);
        data.put("swimmingStyle", this.swimmingStyle);
        data.put("activityDate", this.activityDate);
        data.put("numOfParticipants", this.numOfParticipants);
        data.put("length", this.length);
        data.put("fromAge", this.fromAge);
        data.put("toAge", this.toAge);

        return data;
    }

    public ArrayList<Participant> getParticipants() throws JSONException {
        ArrayList<Participant> participants = new ArrayList<>();
        if (this.participants != null) {
            participants = getParticipants(new JSONObject(this.participants));
        }
        return participants;
    }

    public ArrayList<Participant> getParticipants(JSONObject dataObj) throws JSONException {
        Iterator<String> participantIds = dataObj.keys();
        ArrayList<Participant> participants = new ArrayList<>();

        while (participantIds.hasNext()) {
            String currentId = participantIds.next();
            JSONObject participantJson = new JSONObject(dataObj.get(currentId).toString());
            Participant participant = new Participant(currentId, participantJson);
            participants.add(participant);
        }
        return participants;
    }

    public ArrayList<Participant> getCurrentParticipants() throws JSONException {
        ArrayList<Participant> participants = new ArrayList<>();
        if (this.currentParticipants != null) {
            participants = getParticipants(new JSONObject(this.currentParticipants));
        }
        return participants;
    }

    public void setCurrentParticipants(ArrayList<Participant> participants) throws JSONException {
        this.currentParticipants = getParticipantsStr(participants);
    }

    public void setAllParticipants(ArrayList<Participant> participants) throws JSONException {
        this.participants = getParticipantsStr(participants);
    }

    public String getParticipantsStr(ArrayList<Participant> participants) throws JSONException {
        JSONObject participantsMap = new JSONObject();
        for (Participant participant : participants) {
            JSONObject participantJson = participant.getJSON_Object();
            participantsMap.put(participant.getUid(), participantJson);
        }
        return participantsMap.toString();
    }


    /* GETTERS AND SETTERS */
    public String getSwimmingStyle() {
        return swimmingStyle;
    }

    public void setSwimmingStyle(String swimmingStyle) {
        this.swimmingStyle = swimmingStyle;
    }

    public String getId() {
        return this.id;
    }

    public String getName() {
        return this.name;
    }

    public String getActivityDate() {
        return this.activityDate;
    }

    public int getNumOfParticipants() {
        return Integer.valueOf(this.numOfParticipants);
    }

    public String getLength() {
        return this.length;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setActivityDate(String activityDate) {
        this.activityDate = activityDate;
    }

    public void setNumOfParticipants(Integer numOfParticipants) {
        this.numOfParticipants = String.valueOf(numOfParticipants);
    }

    public void setLength(Integer length) {
        this.length = String.valueOf(length);
    }

    public String getAgesString() {
        return this.fromAge.toString() + " - " + this.toAge.toString();
    }

    public String getFromAge() {
        return this.fromAge;
    }

    public String getToAge() {
        return this.toAge;
    }


    public boolean isCurrentUserRegistered(User currentUser, ArrayList<Participant> participants) {
        for (int i = 0; i < participants.size(); i++) {
            if (currentUser.getUid().equals(participants.get(i).getUid())) {
                return true;
            }
        }
        return false;
    }
}